### PR TITLE
Feature/config location

### PR DESCRIPTION
**Pull request by OrangeBean**
Hi there,

This pull request adds configuration of the gauge-js to package.json to define test names and location

    Tests now searched recursively as other plugins do;
    Add default configuration that define regex to match test file names and location;
    Add optional load of the configuration from package.json;

Now it is possible to specify where test files located, how they named and they don't have to be all in one place.

Default search pattern for test step implementation: `**/Tests/**/*.steps.js`
If you want to override it, please add following section to `package.json` of your test step implementation project:
```
"gauge": {
    "testMatch": ["**/MyTests/**/*.impl.js", "**/SpecialTests/*.special.js", "/some/other/location/*.js"]
  }
```
`package.json` searched traversing up the path from the test project root (`manifest.json`).

This commit should address issue #48.

This not the prettiest code in the universe, but fix the issue. I'm happy to refactor if you guys think it's required.

### CODE REVIEW

**Comment by OrangeBean** [comment]
I broke backward compatibility :( I'll change default pattern to match original behavior.

**Comment by sriv** [comment]
@OrangeBean - thanks for the effort, look forward to your updates!

**Comment by OrangeBean** [comment]
Can you guys help me understand why it fails check? Logs are empty. :( How can I run it on my local before pushing to this pull request?

**Comment by sriv** [comment]
@OrangeBean - I can see the travis logs now, ex: https://travis-ci.org/getgauge-contrib/gauge-js/jobs/313824157

It seems that the functional tests are failing.

**Comment by OrangeBean** [comment]
Can you guys help? There is an issue with html-report plugin.

File: src/config.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+var fs = require("fs");
+var path = require("path");
+
+var PACKAGE_JSON = "package.json";
+
+var defaultConfig = {
+  testMatch: ["**/tests/**/*.js"]
```
Comment (nehashri) [inline src/config.js]: Gauge usually has properties loaded via a properties files which is available to use via environment variables. In this case, this property can be moved to js.properties.

Comment (OrangeBean) [reply src/config.js]: Do you want me to get rid of package.json load and replace it with js.properties load? Or try both sources?

**Comment by OrangeBean** [comment]
Not quite sure what you want me to do with submodule. If I run `git submodule update` on my branch it does nothing.

File: src/config.js (hunk 1 above)
Comment (nehashri) [reply src/config.js]: Try with js.properties load only.

**Comment by nehashri** [comment]
If I look at the files changes, I see that there are changes in the commit hash has of git submodule. If you checkout master, you would see a different commit hash. You probably just have to **not** check in  gauge-proto submodule changes. it should remain in sync with master.

**Comment by OrangeBean** [comment]
I fixed submodule. Didn't have time to look into another issue. Will do it today.

**Comment by nehashri** [comment]
Can you look at adding a few unit tests?

File: src/config.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+function getInstance() {
```
Comment (nehashri) [inline src/config.js]: please add a few unit tests here.

**Comment by nehashri** [comment]
@OrangeBean We also have a CLA in place. Could you please sign it [here](http://gauge-cla.herokuapp.com).

**Comment by OrangeBean** [comment]
Can someone check and retrigger why build failed? It looks completely irrelevant to this change.

**Comment by nehashri** [comment]
@OrangeBean your PR looks good. The tests failed because your branch was out of date with master. However, I still see that there are some changes in master not present in this branch. I am unable to perform an auto rebase with master as there are some conflicts. Can you please rebase your branch with master?

**Comment by OrangeBean** [comment]
Anything else need to be done to make this PR merged and closed?

**Comment by BugDiver** [comment]
@OrangeBean Thanks a lot for your contribution.

**Pull request merged**